man/lvmlockd.8.in
.TH "LVMLOCKD" "8" "LVM TOOLS #VERSION#" "Red Hat, Inc" "\""

.SH NAME
lvmlockd \(em lvm locking daemon

.SH DESCRIPTION
lvm commands use lvmlockd to coordinate access to shared storage.
.br
When lvm is used on devices shared by multiple hosts, locks will:

- coordinate reading and writing of lvm metadata
.br
- validate caching of lvm metadata
.br
- prevent concurrent activation of logical volumes

lvmlockd uses an external lock manager to perform basic locking.
.br
Lock manager (lock type) options are:

- sanlock: places locks on disk within lvm storage.
.br
- dlm: uses network communication and a cluster manager.

.SH OPTIONS

lvmlockd [options]

For default settings, see lvmlockd -h.

.B  --help | -h
        Show this help information.

.B  --version | -V
        Show version of lvmlockd.

.B  --test | -T
        Test mode, do not call lock manager.

.B  --foreground | -f
        Don't fork.

.B  --daemon-debug | -D
        Don't fork and print debugging to stdout.

.B  --pid-file | -p
.I path
        Set path to the pid file.

.B  --socket-path | -s
.I path
        Set path to the socket to listen on.

.B  --local-also | -a
        Manage locks between pids for local vgs.

.B  --local-only | -o
        Only manage locks for local vgs, not dlm|sanlock vgs.

.B  --gl-type | -g
.I str
        Set global lock type to be dlm|sanlock.

.B  --system-id | -y
.I str
        Set the local system id.

.B  --host-id | -i
.I num
        Set the local sanlock host id.


.SH USAGE

.SS Initial set up

Using lvm with lvmlockd for the first time includes some one-time set up
steps:

.SS 1. choose a lock manager

.I dlm
.br
If dlm (or corosync) are already being used by other cluster
software, then select dlm.  dlm uses corosync which requires additional
configuration beyond the scope of this document.  See corosync and dlm
documentation for instructions on configuration, setup and usage.

.I sanlock
.br
Choose sanlock if dlm/corosync are not otherwise required.
sanlock does not depend on any clustering software or configuration.
On each host using lvmlockd, set lvm.conf lvmlockd_sanlock_host_id to a
unique integer (1-2000).

.SS 2. configure hosts to use lvmlockd

On all hosts running lvmlockd, configure lvm.conf:
.nf
locking_type = 1
use_lvmetad = 1
use_lvmlockd = 1
.fi

(Start lvmetad if it was not already running.)

.SS 3. start lvmlockd

Use a service/init file if available, or just run "lvmlockd".

.SS 4. start lock manager

.I sanlock
.br
systemctl start wdmd sanlock

.I dlm
.br
Follow external clustering documentation when applicable, otherwise:
.br
systemctl start corosync dlm

.SS 5. create vgs on shared devices

vgcreate --lock-type sanlock|dlm <vg_name> <devices>

The vgcreate --lock-type option means that lvm commands will perform
locking for the vg using lvmlockd and the specified lock manager.

.SS 6. start vgs on all hosts

vgchange --lock-start

lvmlockd includes the concept of "starting" the locking for vgs that
have been created with a lock type.  Starting the locking can take some
time, and until the start completes, locks are not available.  Reading and
reporting lvm commands are allowed while start is in progress.

.SS 7. create and activate lvs

An lv activated exclusively on one host cannot be activated on another.
When multiple hosts need to use the same lv concurrently, the lv can be
activated with a shared lock (see lvchange options -aey vs -asy.)
(Shared locks are disallowed for lv types that cannot be used from
multiple hosts.)

.SS Subsequent start up

.nf
After initial set up, start up includes:

- start lvmetad
- start lvmlockd
- start lock manager
- vgchange --lock-start
- activate lvs

The shut down sequence is the reverse:

- deactivate lvs
- vgchange --lock-stop
- stop lock manager
- stop lvmlockd
- stop lvmetad
.fi


.SH TOPICS

.SS locking terms

The following terms are used to distinguish vgs that involve locking from
those that do not:

.I "dlock vg"

A "dlock vg" refers to a vg that has a "lock type" of dlm or sanlock.
dlock vgs exist on shared storage that is visible to more than one host.
lvm commands and lvmlockd perform locking for these vgs when they are
used.

If the lock manager for a given lock type is not available (e.g. not
started or has failed), lvmlockd will not be able to acquire locks, and
lvm commands will not be able to fully use dlock vgs.  lvm commands will
typically allow reading and reporting in this condition, but changes and
activation are not allowed.  Maintaining a properly running lock manager
can require skills not covered in this document.

.I "local vg"

A "local vg" refers to a vg that has no lock type or has lock type "none".
lvm commands and lvmlockd do not perform locking for these vgs.  A local
vg typically exists on local (non-shared) devices and cannot be used
concurrently from different hosts.

If a local vg does exist on shared devices, it should have its "system id"
set.  Only the host with a matching system id can then use the local vg.
A vg with no lock type and no system id should be excluded from all but
one host using lvm.conf filters.  Without any of these protections, a
local vg on shared devices can be easily damaged or destroyed.

(When lvmlockd is enabled, it actively manages locks for dlock vgs, but
also keeps a record of local vgs so that it can quickly determine that no
locks are needed for a given local vg.)

.I "clvm vg"

A "clvm vg" refers to a vg with the "clustered" flag set by a host using
clvmd.  These vgs cannot be used by hosts using lvmlockd, only by hosts
using clvmd.  These vgs have the CLUSTERED status flag in the vg metadata,
and will have no lock type set.  See below for converting a clvm vg to a
dlock vg.

The term "clustered" is widely used in lvm documentation, and refers to
clvm vgs.  Statements about "clustered" vgs to not apply to dlock vgs.  A
new set of rules, properties and descriptions apply to dlock vgs, created
with a "lock type", as opposed to clvm vgs, created with the "clustered"
flag.


.SS locking activity

To optimize the use of lvm with lvmlockd, consider the three kinds of lvm
locks and when they are used:

1.
.I gl lock

The global lock (gl) is associated with global information, which is
information not isolated to a single vg.  This is primarily:

.nf
- the list of all vg names
- the list of pvs not allocated to a vg (orphan pvs)
- properties of orphan pvs, e.g. pv size
.fi

The global lock is used in shared mode by commands that want to read this
information, or in exclusive mode by commands that want to change this
information.

"vgs" acquires the global lock in shared mode because it reports the list
of all vg names.

"vgcreate" acquires the global lock in exclusive mode because it creates a
new vg name, and it takes a pv from the list of unused pvs.

When use_lvmlockd is enabled, lvm commands attempt to acquire the global
lock even if no dlock vgs exist, so lvmlockd should not be enabled unless
dlock vgs will be used.

2.
.I vg lock

A vg lock is associated with each vg.  The vg lock is acquired in shared
mode to read the vg and in exclusive mode to change the vg (write the vg
metadata).  This serializes modifications to a vg with all other lvm
commands on the vg.

"vgs" will not only acquire the global lock (see above), but will acquire
the vg lock for each vg prior to reading it.

"vgs vg_name" does not acquire the global lock (it does not need the list
of all vg names), but will acquire the vg lock on each vg_name listed.

3.
.I lv lock

An lv lock is acquired before the lv is activated, and is released after
the lv is deactivated.  If the lv lock cannot be acquired, the lv is not
activated.  lv locks are persistent and remain in place after the
activation command is done.  gl and vg locks are transient, and are held
only while an lvm command is running.


Reporting commands can sometimes lead to unexpected and excessive locking
activity.  See below for optimizing reporting commands to avoid unwanted
or intrusive locking.


lib/locking/lvmlockd.c
/*
 * Copyright (C) 2013 Red Hat, Inc.
 *
 * This file is part of LVM2.
 *
 * This copyrighted material is made available to anyone wishing to use,
 * modify, copy, or redistribute it subject to the terms and conditions
 * of the GNU Lesser General Public License v.2.1.
 */

#include "lib.h"
#include "toolcontext.h"
#include "metadata.h"
#include "segtype.h"
#include "lvmetad.h"
#include "lvmlockd.h"
#include "lvmcache.h"
#include "lvmlockd-client.h"

static daemon_handle _lvmlockd;
static int _lvmlockd_active;
static int _lvmlockd_connected;

static const char *_lvmlockd_socket = NULL;
static struct cmd_context *_lvmlockd_cmd = NULL;

/* The name of the internal lv created to hold sanlock locks. */
#define SANLOCK_LV_NAME "lvmlock"

void lvmlockd_disconnect(void)
{
	if (_lvmlockd_connected)
		daemon_close(_lvmlockd);
	_lvmlockd_connected = 0;
	_lvmlockd_cmd = NULL;
}

void lvmlockd_init(struct cmd_context *cmd)
{
	if (!_lvmlockd_active && !access(LVMLOCKD_PIDFILE, F_OK))
		log_warn("WARNING: lvmlockd is running but disabled.");
	if (!_lvmlockd_active)
		return;
	_lvmlockd_cmd = cmd;
}

static void _lvmlockd_connect(void)
{
	if (!_lvmlockd_active || !_lvmlockd_socket || _lvmlockd_connected)
		return;

	_lvmlockd = lvmlockd_open(_lvmlockd_socket);

	if (_lvmlockd.socket_fd >= 0 && !_lvmlockd.error) {
		log_debug("Successfully connected to lvmlockd on fd %d.",
			  _lvmlockd.socket_fd);
		_lvmlockd_connected = 1;
	}
}

void lvmlockd_connect_or_warn(void)
{
	if (!_lvmlockd_active || _lvmlockd_connected)
		return;

	_lvmlockd_connect();

	if (!_lvmlockd_connected) {
		log_warn("WARNING: Failed to connect to lvmlockd: %s. No distributed locks available.",
			 strerror(_lvmlockd.error));
	}
}

/*
 * in command setup:
 *
 * 1. if use_lvmlockd is set in config,
 *    lvmlockd_set_active() sets _lvmlockd_active = 1
 *
 * 2. lvmlockd_init() sees _lvmlockd_active, and sets _lvmlockd_cmd
 *
 * 3. lvmlockd_connect_or_warn()/_lvmlockd_connect() see _lvmlockd_active,
 *    create connection and if successful set _lvmlockd_connected = 1
 *
 * in command processing:
 *
 * 1. dlock function calls lvmlockd_connected() which returns
 *    _lvmlockd_connected
 *
 * 2. if lvmlockd_connected() returns 0, dlock function fails
 */

static int lvmlockd_connected(void)
{
	if (_lvmlockd_connected)
		return 1;

	log_error("Error: lvmlockd connection does not exist.");
	return 0;
}

void lvmlockd_set_active(int active)
{
	_lvmlockd_active = active;
}

void lvmlockd_set_socket(const char *sock)
{
	_lvmlockd_socket = sock;
}

static daemon_reply _lvmlockd_send(const char *req_name, ...)
{
	va_list ap;
	daemon_reply repl;
	daemon_request req;

	req = daemon_request_make(req_name);

	va_start(ap, req_name);
	daemon_request_extend_v(req, ap);
	va_end(ap);

	repl = daemon_send(_lvmlockd, req);

	daemon_request_destroy(req);

	return repl;
}

static void result_str_to_flags(const char *str, uint32_t *flags)
{
	if (strstr(str, "LOCAL_VG"))
		*flags |= LD_RF_LOCAL_VG;

	if (strstr(str, "NO_LOCKSPACES"))
		*flags |= LD_RF_NO_LOCKSPACES;

	if (strstr(str, "NO_GL_LS"))
		*flags |= LD_RF_NO_GL_LS;
}

/*
 * evaluate the reply from lvmlockd, check for errors, extract
 * the result and result_flags returned by lvmlockd.
 * 0 failure (no result/result_flags set)
 * 1 success (result/result_flags set)
 */

static int _lvmlockd_result(daemon_reply reply, int *result, uint32_t *result_flags)
{
	int reply_result;
	const char *reply_flags;

	if (reply.error) {
		log_error("lvmlockd_result reply error %d", reply.error);
		return 0;
	}

	if (strcmp(daemon_reply_str(reply, "response", ""), "OK")) {
		log_error("lvmlockd_result bad response");
		return 0;
	}

	/* FIXME: using -1000 is dumb */

	reply_result = daemon_reply_int(reply, "op_result", -1000);
	if (reply_result == -1000) {
		log_error("lvmlockd_result no op_result");
		return 0;
	}

	log_debug("lvmlockd_result %d", reply_result);

	*result = reply_result;

	if (!result_flags)
		goto out;

	reply_flags = daemon_reply_str(reply, "result_flags", NULL);
	if (reply_flags) {
		log_debug("lvmlockd_result flags %s", reply_flags);
		result_str_to_flags(reply_flags, result_flags);
	}
 out:
	return 1;
}

/*
 * Check if a lock_type uses lvmlockd.
 * If not (none, clvm), return 0.
 * If so, (dlm sanlock), return LOCK_TYPE_
 */

int dlock_type(const char *lock_type)
{
	if (!lock_type)
		return 0;

	if (!strcmp(lock_type, "dlm"))
		return LOCK_TYPE_DLM;
	if (!strcmp(lock_type, "sanlock"))
		return LOCK_TYPE_SANLOCK;

	return 0;
}

/*
 * result/result_flags are values returned from lvmlockd.
 *
 * return 0 (failure)
 * return 1 (result/result_flags indicate success/failure)
 *
 * return 1 result 0   (success)
 * return 1 result < 0 (failure)
 *
 * caller may ignore result < 0 failure depending on
 * result_flags and the specific command/mode.
 *
 * When this function returns 0 (failure), no result/result_flags
 * were obtained from lvmlockd.
 *
 * When this function returns 1 (success), result/result_flags may
 * have been obtained from lvmlockd.  This lvmlockd result may
 * indicate a locking failure.
 */

int dlock_general(struct cmd_context *cmd,
		  const char *cmd_name,
		  const char *req_name,
		  const char *vg_name,
		  const char *vg_lock_type,
		  const char *vg_lock_args,
		  const char *lv_name,
		  const char *lv_lock_args,
		  const char *mode,
		  const char *opts,
		  int *result,
		  uint32_t *result_flags)
{
	daemon_reply reply;
	int pid = getpid();

	*result = 0;
	*result_flags = 0;

	if (vg_lock_type && !dlock_type(vg_lock_type))
		return 1;

	if (!strcmp(mode, "na"))
		return 1;

	if (!_lvmlockd_active)
		return 1;
	if (!lvmlockd_connected())
		return 0;

	if (!opts)
		opts = "none";

	/* cmd and pid are passed for informational and debugging purposes */

	if (vg_name && lv_name) {
		reply = _lvmlockd_send(req_name,
					"cmd = %s", cmd_name,
					"pid = %d", pid,
					"mode = %s", mode,
					"opts = %s", opts,
					"vg_name = %s", vg_name,
					"lv_name = %s", lv_name,
					"vg_lock_type = %s", vg_lock_type ?: "none",
					"vg_lock_args = %s", vg_lock_args ?: "none",
					"lv_lock_args = %s", lv_lock_args ?: "none",
					NULL);

		if (!_lvmlockd_result(reply, result, result_flags))
			goto fail;

		log_debug("dlock %s %s vg %s lv %s result %d %x",
			  req_name, mode, vg_name, lv_name, *result, *result_flags);

	} else if (vg_name) {
		reply = _lvmlockd_send(req_name,
					"cmd = %s", cmd_name,
					"pid = %d", pid,
					"mode = %s", mode,
					"opts = %s", opts,
					"vg_name = %s", vg_name,
					"vg_lock_type = %s", vg_lock_type ?: "none",
					"vg_lock_args = %s", vg_lock_args ?: "none",
					NULL);

		if (!_lvmlockd_result(reply, result, result_flags))
			goto fail;

		log_debug("dlock %s %s vg %s result %d %x",
			  req_name, mode, vg_name, *result, *result_flags);

	} else {
		reply = _lvmlockd_send(req_name,
					"cmd = %s", cmd_name,
					"pid = %d", pid,
					"mode = %s", mode,
					"opts = %s", opts,
					"vg_lock_type = %s", vg_lock_type ?: "none",
					NULL);

		if (!_lvmlockd_result(reply, result, result_flags))
			goto fail;

		log_debug("dlock %s %s result %d %x",
			  req_name, mode, *result, *result_flags);
	}

	daemon_reply_destroy(reply);

	/* result/result_flags have lvmlockd result */
	return 1;

 fail:
	/* no result was obtained from lvmlockd */

	log_error("dlock %s %s failed no lvmlockd result", req_name, mode);

	daemon_reply_destroy(reply);
	return 0;
}

/*
 * Called after a vg_write to indicate that the version number
 * of the vg lock should be incremented (set to the new vg seqno)
 * to invalidate the vg cache on other hosts.
 */

int dlock_vg_update(struct volume_group *vg)
{
	daemon_reply reply;
	int result;
	int ret;

	if (!dlock_type(vg->lock_type))
		return 1;

	if (!_lvmlockd_active)
		return 1;
	if (!lvmlockd_connected())
		return 0;

	log_debug("dlock update %s %u", vg->name, vg->seqno);

	reply = _lvmlockd_send("vg_update",
				"vg_name = %s", vg->name,
				"version = %d", (int64_t)vg->seqno,
				NULL);

	if (!_lvmlockd_result(reply, &result, NULL)) {
		ret = 0;
	} else {
		ret = (result < 0) ? 0 : 1;
	}

	daemon_reply_destroy(reply);
	return ret;
}

static struct logical_volume *find_sanlock_lv(struct volume_group *vg,
					      const char *lock_lv_name)
{
	struct lv_list *lvl;

	dm_list_iterate_items(lvl, &vg->lvs) {
		if (!strcmp(lvl->lv->name, lock_lv_name))
			return lvl->lv;
	}
	return NULL;
}

/*
 * Eventually we may create a separate lvmlocks lv on multiple pvs.
 * They will not be mirrored, but will be independent lvs, and
 * sanlock will use all of them.  sanlock will be able to continue locking as
 * long as the majority of them are still available.
 */

static int create_sanlock_lv(struct cmd_context *cmd, struct volume_group *vg,
			      const char *lock_lv_name)
{
	struct logical_volume *lv;
	struct lvcreate_params lp = {
		.activate = CHANGE_ALY,
		.alloc = ALLOC_INHERIT,
		.extents = 1073741824 / vg->extent_size, /* -L 1G */
		.major = -1,
		.minor = -1,
		.permission = LVM_READ | LVM_WRITE,
		.pvh = &vg->pvs,
		.read_ahead = DM_READ_AHEAD_NONE,
		.stripes = 1,
		.vg_name = vg->name,
		.lv_name = dm_pool_strdup(cmd->mem, lock_lv_name),
		.zero = 1,
	};

	dm_list_init(&lp.tags);

	if (!(lp.segtype = get_segtype_from_string(vg->cmd, "striped")))
		return_0;

	/* TODO: configurable default size? */

	lv = lv_create_single(vg, &lp);
	if (!lv) {
		log_error("Failed to create sanlock lv %s in vg %s", lock_lv_name, vg->name);
		return 0;
	}

	lv_set_hidden(lv);
	return 1;
}

/*
 * vg_lock_args format for sanlock is
 * version_string:undefined:lock_lv_name
 *
 * vg_lock_args format for dlm is
 * version_string:undefined:cluster_name
 *
 * lv_lock_args format for sanlock is
 * version_string:undefined:offset
 *
 * lv_lock_args is not used for dlm
 *
 * undefined may contain ":"
 */

static const char *vg_lock_args_to_lv(const char *vg_lock_args)
{
	const char *args = vg_lock_args;
	const char *colon, *last = NULL;

	while (1) {
		if (!args || (*args == '\0'))
			break;
		colon = strstr(args, ":");
		if (!colon)
			break;
		last = colon;
		args = colon + 1;
	}

	if (last)
		return last + 1;
	else
		return NULL;
}

static int remove_sanlock_lv(struct cmd_context *cmd, struct volume_group *vg,
			     const char *lock_lv_name)
{
	struct logical_volume *lv;

	lv = find_sanlock_lv(vg, lock_lv_name);
	if (!lv) {
		log_error("Failed to find sanlock lv %s in vg %s", lock_lv_name, vg->name);
		return 0;
	}

	if (!lv_remove(lv)) {
		log_error("Failed to remove sanlock lv %s/%s", vg->name, lock_lv_name);
		return 0;
	}

	return 1;
}

static int activate_sanlock_lv(struct cmd_context *cmd, struct volume_group *vg)
{
	struct logical_volume *lv;
	const char *lock_lv_name = vg_lock_args_to_lv(vg->lock_args);

	lv = find_sanlock_lv(vg, lock_lv_name);
	if (!lv) {
		log_error("Failed to find sanlock lv %s in vg %s", lock_lv_name, vg->name);
		return 0;
	}

	if (!activate_lv(cmd, lv)) {
		log_error("Failed to activate sanlock lv %s/%s", vg->name, lock_lv_name);
		return 0;
	}

	return 1;
}

static int deactivate_sanlock_lv(struct cmd_context *cmd, struct volume_group *vg)
{
	struct logical_volume *lv;
	const char *lock_lv_name = vg_lock_args_to_lv(vg->lock_args);

	lv = find_sanlock_lv(vg, lock_lv_name);
	if (!lv) {
		log_error("Failed to find sanlock lv %s in vg %s", lock_lv_name, vg->name);
		return 0;
	}

	if (!deactivate_lv(cmd, lv)) {
		log_error("Failed to deactivate sanlock lv %s/%s", vg->name, lock_lv_name);
		return 0;
	}

	return 1;
}

/*
 * Starting a vg involves:
 * 1. reading the vg without a lock
 * 2. getting the lock_type/lock_args from the vg
 * 3. doing start_vg in lvmlockd for the lock_type 
 *    this means joining the lockspace
 *
 * The vg read in step 1 should not be used for anything
 * other than getting the lock_type/lock_args/uuid necessary
 * for starting the lockspace.  To use the vg for anything
 * else would require dlock_vg() and then rereading the
 * vg to ensure a valid copy is used.
 */

int dlock_start_vg(struct cmd_context *cmd, struct volume_group *vg,
		   const char *cmd_mode)
{
	char uuid[64] __attribute__((aligned(8)));
	daemon_reply reply;
	const char *lock_type;
	int host_id = 0;
	int result;
	int ret;

	memset(uuid, 0, sizeof(uuid));

	/* We may want to ignore non-dlock vgs here and
	   inform lvmlockd of local vgs another way, or
	   have it discover local vgs itself. */

	/*
	if (!dlock_type(vg->lock_type))
		return 1;
	*/

	if (!_lvmlockd_active)
		return 1;
	if (!lvmlockd_connected())
		return 0;

	if (cmd_mode && !strcmp(cmd_mode, "na"))
		return 1;

	log_debug("dlock_start_vg %s lock_type %s", vg->name,
		  vg->lock_type ? vg->lock_type : "empty");

	if (vg->lock_type && !strcmp(vg->lock_type, "sanlock")) {
		host_id = find_config_tree_int(cmd, global_lvmlockd_sanlock_host_id_CFG, NULL);
		if (host_id < 1 || host_id > 2000) {
			log_error("Invalid sanlock host_id.");
			return 0;
		}

		/*
		 * This is a crucial difference between starting
		 * sanlock vs dlm vgs: the internal sanlock lv
		 * needs to be activated before lvmlockd does the
		 * start because sanlock needs to use the lv.
		 */
		if (!activate_sanlock_lv(cmd, vg))
			return 0;
	}

	id_write_format(&vg->id, uuid, sizeof(uuid));

	if (!dlock_type(vg->lock_type)) {
		char *sysid = NULL;

		if (vg->system_id && (strlen(vg->system_id) > 0))
			sysid = vg->system_id;

		reply = _lvmlockd_send("add_local",
				"pid = %d", getpid(),
				"vg_name = %s", vg->name,
				"vg_uuid = %s", uuid[0] ? uuid : "none",
				"vg_sysid = %s", sysid ?: "none",
				NULL);

		lock_type = "local";
	} else {
		reply = _lvmlockd_send("start_vg",
				"pid = %d", getpid(),
				"vg_name = %s", vg->name,
				"vg_lock_type = %s", vg->lock_type,
				"vg_lock_args = %s", vg->lock_args,
				"vg_uuid = %s", uuid[0] ? uuid : "none",
				"host_id = %d", (int64_t)host_id,
				NULL);

		lock_type = vg->lock_type;
	}

	if (!_lvmlockd_result(reply, &result, NULL)) {
		result = -1;
		ret = 0;
	} else {
		ret = (result < 0) ? 0 : 1;
	}

	if (!ret)
		log_error("Locking start %s VG %s %d", lock_type, vg->name, result);
	else
		log_debug("dlock_start_vg %s done", vg->name);

	daemon_reply_destroy(reply);

	return ret;
}

int dlock_stop_vg(struct cmd_context *cmd, struct volume_group *vg)
{
	daemon_reply reply;
	int result;
	int ret;

	if (!dlock_type(vg->lock_type))
		return 1;

	log_debug("dlock_stop_vg %s lock_type %s", vg->name,
		  vg->lock_type ? vg->lock_type : "empty");

	if (!_lvmlockd_active)
		return 1;
	if (!lvmlockd_connected())
		return 0;

	reply = _lvmlockd_send("stop_vg",
				"pid = %d", getpid(),
				"vg_name = %s", vg->name,
				NULL);

	if (!_lvmlockd_result(reply, &result, NULL)) {
		ret = 0;
	} else {
		ret = (result < 0) ? 0 : 1;
	}

	if (!ret) {
		log_error("Locking stop %s VG %s %d", vg->lock_type, vg->name, result);
		goto out;
	}

	if (!strcmp(vg->lock_type, "sanlock")) {
		log_debug("dlock_stop_vg deactivate sanlock lv");
		deactivate_sanlock_lv(cmd, vg);
	}
out:
	daemon_reply_destroy(reply);

	return ret;
}

/*
 * vgcreate does:
 * dlock_init_vg();
 * vg_create
 * dlock_vg_update();
 * dlock_start_vg();
 *
 * TODO:
 * The update will not work because it's before the start.
 * Avoid reporting an error in that case, and perhaps save
 * the version and have start_vg acquire the vg lock to
 * save the version.
 */

int dlock_init_vg_sanlock(struct cmd_context *cmd, struct volume_group *vg)
{
	daemon_reply reply;
	const char *reply_str;
	const char *vg_lock_args = NULL;
	const char *lock_lv_name = SANLOCK_LV_NAME;
	int result;
	int ret;

	log_debug("dlock_init_vg_sanlock %s", vg->name);

	if (!_lvmlockd_active)
		return 1;
	if (!lvmlockd_connected())
		return 0;

	if (!create_sanlock_lv(cmd, vg, lock_lv_name)) {
		log_error("Failed to create internal lv.");
		return 0;
	}

	/*
	 * N.B. this passes the lock_lv_name as vg_lock_args
	 * even though it is only part of the final args string
	 * which will be returned from lvmlockd.
	 */

	reply = _lvmlockd_send("init_vg",
				"pid = %d", getpid(),
				"vg_name = %s", vg->name,
				"vg_lock_type = %s", "sanlock",
				"vg_lock_args = %s", lock_lv_name,
				NULL);

	if (!_lvmlockd_result(reply, &result, NULL)) {
		ret = 0;
	} else {
		ret = (result < 0) ? 0 : 1;
	}

	if (!ret) {
		log_error("dlock_init_vg_sanlock lvmlockd result %d", result);
		remove_sanlock_lv(cmd, vg, lock_lv_name);
		goto out;
	}

	reply_str = daemon_reply_str(reply, "vg_lock_args", NULL);
	if (!reply_str) {
		log_error("vg_lock_args not returned");
		ret = 0;
		goto out;
	}

	vg_lock_args = dm_pool_strdup(cmd->mem, reply_str);
	if (!vg_lock_args) {
		log_error("vg_lock_args allocation failed");
		ret = 0;
        }
 out:
	daemon_reply_destroy(reply);

	vg->lock_args = vg_lock_args;
	return ret;
}

int dlock_init_vg_dlm(struct cmd_context *cmd, struct volume_group *vg)
{
	daemon_reply reply;
	const char *reply_str;
	const char *vg_lock_args = NULL;
	int result;
	int ret;

	log_debug("dlock_init_vg_dlm %s", vg->name);

	if (!_lvmlockd_active)
		return 1;
	if (!lvmlockd_connected())
		return 0;

	reply = _lvmlockd_send("init_vg",
				"pid = %d", getpid(),
				"vg_name = %s", vg->name,
				"vg_lock_type = %s", "dlm",
				NULL);

	if (!_lvmlockd_result(reply, &result, NULL)) {
		ret = 0;
	} else {
		ret = (result < 0) ? 0 : 1;
	}

	if (!ret) {
		log_error("dlock_init_vg_dlm lvmlockd result %d", result);
		goto out;
	}

	reply_str = daemon_reply_str(reply, "vg_lock_args", NULL);
	if (!reply_str) {
		log_error("vg_lock_args not returned");
		ret = 0;
		goto out;
	}

	vg_lock_args = dm_pool_strdup(cmd->mem, reply_str);
	if (!vg_lock_args) {
		log_error("vg_lock_args allocation failed");
		ret = 0;
	}
 out:
	daemon_reply_destroy(reply);

	vg->lock_args = vg_lock_args;
	return ret;
}

/*
 * For lvcreate, sanlock needs to allocate/initialize a lease area
 * on the internal leases lv to use for the lock for the new lv.
 */

int dlock_init_lv_sanlock(struct cmd_context *cmd,
			  struct volume_group *vg, const char *lv_name,
			  const char *lock_type, const char **lock_args_ret)
{
	daemon_reply reply;
	char *lv_lock_args = NULL;
	const char *reply_str;
	int result;
	int ret;

	log_debug("dlock_init_lv_sanlock %s/%s", vg->name, lv_name);

	if (!_lvmlockd_active)
		return 1;
	if (!lvmlockd_connected())
		return 0;

	reply = _lvmlockd_send("init_lv",
				"pid = %d", getpid(),
				"vg_name = %s", vg->name,
				"lv_name = %s", lv_name,
				"vg_lock_type = %s", vg->lock_type,
				"vg_lock_args = %s", vg->lock_args,
				NULL);

	if (!_lvmlockd_result(reply, &result, NULL)) {
		ret = 0;
	} else {
		ret = (result < 0) ? 0 : 1;
	}

	if (!ret) {
		log_error("dlock_init_lv_sanlock lvmlockd result %d", result);
		goto out;
	}

	reply_str = daemon_reply_str(reply, "lv_lock_args", NULL);
	if (!reply_str) {
		log_error("lv_lock_args not returned");
		ret = 0;
		goto out;
	}

	lv_lock_args = dm_pool_strdup(cmd->mem, reply_str);
	if (!lv_lock_args) {
		log_error("lv_lock_args allocation failed");
		ret = 0;
	}
 out:
	daemon_reply_destroy(reply);

	*lock_args_ret = lv_lock_args;
	return ret;
}

/*
 * vgremove for sanlock:
 *
 * lv remove any existing lv's:
 * lock each lv ex, unlock rename and erase each
 * (To simplify things initially, perhaps just fail if all
 * lvs have not yet been removed.)
 *
 * lock vg ex, unlock rename to "vg_remove"
 *
 * Any other hosts are joined to the lockspace need to stop
 * the vg before we can remove and erase the sanlock lv.
 * lvmlockd can query for other live hosts in the lockspace,
 * and could use a sanlock host_message to tell them to leave.
 * We could wait for some timeout for all other hosts to leave
 * and then fail the vgremove if they haven't.
 *
 * When other hosts are gone, leave the sanlock lockspace ourself,
 * lvremove the sanlock lv, and continue with the standard vgremove.
 */

/* called before vg_remove on disk */
int dlock_free_vg_sanlock(struct cmd_context *cmd, struct volume_group *vg)
{
	daemon_reply reply;
	int result;
	int ret;

	/* Unlocking the vg lock here preempts the dlock_vg("un") in
	   toollib.c which will occur after the lockspace is stopped. */

	log_debug("dlock_free_vg_sanlock for vgremove %s", vg->name);

	if (!_lvmlockd_active)
		return 1;
	if (!lvmlockd_connected())
		return 0;

	/*
	 * lvmlockd could combine these steps in a "free_vg" command:
	 * dlock_vg unlock-rename, dlock_gl unlock-rename,
	 * check for other hosts in lockspace, notify them to leave,
	 * wait for all to be gone, leave the lockspace.
	 * Or, we could do it as separate steps from here.
	 */

	reply = _lvmlockd_send("free_vg",
				"pid = %d", getpid(),
				"vg_name = %s", vg->name,
				"vg_lock_type = %s", vg->lock_type,
				"vg_lock_args = %s", vg->lock_args,
				NULL);

	if (!_lvmlockd_result(reply, &result, NULL)) {
		ret = 0;
	} else {
		ret = (result < 0) ? 0 : 1;
	}

	if (!ret) {
		log_error("dlock_free_vg_sanlock lvmlockd result %d", result);
		goto out;
	}

	deactivate_sanlock_lv(cmd, vg);

	remove_sanlock_lv(cmd, vg, vg_lock_args_to_lv(vg->lock_args));
 out:
	daemon_reply_destroy(reply);

	return ret;
}

/* called after vg_remove on disk */
int dlock_free_vg_dlm(struct cmd_context *cmd, struct volume_group *vg)
{
	uint32_t result_flags;
	int result;
	int ret;

	/* Unlocking the vg lock here preempts the dlock_vg("un") in
	   toollib.c which will occur after the lockspace is stopped. */

	log_debug("dlock_free_vg_dlm un for vgremove %s", vg->name);

	/* dlock_vg(cmd, vg->name, "un", 0); */

	ret = dlock_general(cmd, "vgremove", "lock_vg", vg->name,
			    NULL, NULL, NULL, NULL, "un", NULL,
			    &result, &result_flags);

	if (!ret || result < 0) {
		log_error("dlock_free_vg_dlm lvmlockd result %d", result);
		return 0;
	}

	/* TODO: notify other hosts to leave this lockspace? */

	dlock_stop_vg(cmd, vg);

	return 1;
}

/*
 * After lvremove has removed the lv, this is called to let
 * sanlock clear the lease area it had used for the lv.
 */

int dlock_free_lv_sanlock(struct cmd_context *cmd,
			  struct volume_group *vg, const char *lv_name,
			  const char *lock_type, const char *lock_args)
{
	daemon_reply reply;
	int result;
	int ret;

	log_debug("dlock_free_lv_sanlock for lvremove %s/%s",
		  vg->name, lv_name);

	if (!_lvmlockd_active)
		return 1;
	if (!lvmlockd_connected())
		return 0;

	/* TODO: lv_lock_type should be removed */

	reply = _lvmlockd_send("free_lv",
				"pid = %d", getpid(),
				"vg_name = %s", vg->name,
				"lv_name = %s", lv_name,
				"vg_lock_type = %s", vg->lock_type,
				"vg_lock_args = %s", vg->lock_args,
				"lv_lock_type = %s", lock_type,
				"lv_lock_args = %s", lock_args,
				NULL);

	if (!_lvmlockd_result(reply, &result, NULL)) {
		ret = 0;
	} else {
		ret = (result < 0) ? 0 : 1;
	}
	
	if (!ret) {
		log_error("dlock_free_lv_sanlock lvmlockd result %d", result);
	}

	daemon_reply_destroy(reply);

	return ret;
}

/* For now, we will not allow vg rename of a dlock_type vg. */

#if 0
int dlock_rename_vg(struct volume_group *vg, const char *vg_name)
{
	/* Perhaps require a dlock_type vg to be converted to a
	 * local vg before being renamed, then converted back to
	 * its original type? */

	/*
	 * In lvmlockd, roughly:
	 *
	 * dlock vg ex
	 * dlock vg un (unlock destroy vglk)
	 *
	 * check/fail if any other nodes hold any lv locks
	 * check/fail if any local non-persisent lv locks exist
	 * unlock any local persistent lv locks
	 *
	 * sanlock:
	 * rem_lockspace(vg->name);
	 * sanlock_init(ls_vg_name);
	 * sanlock_init(res_vglk);
	 * add_lockspace(vg_name);
	 * dlm:
	 * release_lockspace
	 * new_lockspace
	 *
	 * dlock vg ex
	 * reacquire any local persistent lv locks
	 */
}
#endif

